src/ex2/CompileVisitor.java

```diff
@@ -11,286 +11,329 @@
+		StringBuilder sig = new StringBuilder(); // signature
+		StringBuilder callArgs = new StringBuilder(); // call arguments for next command
+		String returnType = TypeDecider.llvmType(funcSymbol.getDecl().split(Symbol.DECL_MAJOR_SEP)[1]);
+		sig.append(returnType);
+		sig.append(" (i8*");
+		String[] splitted = funcSymbol.getDecl().split(Symbol.DECL_MAJOR_SEP)[0].split(Symbol.DECL_SEP);
+		for (int i = 0; i < splitted.length; i++) {
+			sig.append(", ");
+			callArgs.append(", ");
+			sig.append(TypeDecider.llvmType(splitted[i]));
+			callArgs.append(TypeDecider.llvmType(splitted[i]));
+			callArgs.append(" %_").append(actualsRegisters.indexOf(i));
+		}
+		sig.append(")*");
+		s = String.format("%%_%d = bitcast i8* %%_%d to %s", castedFunc, funcPtr, sig.toString());
+		addLine(s);
+
+		int resultPtr = lastRegisterNumber++; // calling the function with args that were created at the top
+		s = String.format("%%_%d = call %s %%_%d(i8* %%_%d%s)", resultPtr, returnType, castedFunc, ownerPtr,
+				callArgs.toString());
+		addLine(s);
+	}
+
+	@Override
+	public void visit(ThisExpr e) {
+		String s = String.format("%%_%d = add i8* 0, %%this", lastRegisterNumber++);
+		addLine(s);
+	}
+
+	@Override
+	public void visit(NewObjectExpr e) {
+		// allocation
+		int tmp0 = lastRegisterNumber++;
+		String s = String.format("%%_%d = call i8* @calloc(i32 1, i32 %d)", tmp0,
+				pointerMap.get(e.classId()).getClassSize());
+		addLine(s);
+
+		// setting the vtable
+		int tmp1 = lastRegisterNumber++;
+		s = String.format("%%_%d = bitcast i8* %%_%d to i8***", tmp1, tmp0);
+		addLine(s);
+		int tmp2 = lastRegisterNumber++;
+		int classFuncs = pointerMap.get(e.classId()).numberOfFuncs();
+		s = String.format("%%_%d = getelementptr [%d x i8*], [%d x i8*]* @.%s_vtable, i32 0, i32 0", tmp2, classFuncs,
+				classFuncs, e.classId());
+		addLine(s);
+		s = String.format("store i8** %%_%d, i8*** %%_%d", tmp2, tmp1);
+		addLine(s);
+		// creating the result which can be saved
+		int resReg = lastRegisterNumber++;
+		s = String.format("%%_%d = add i8* 0, %d", resReg, tmp0);
+		addLine(s);
+	}
+
+	@Override
+	public void visit(FormalArg formalArg) {
+		String s = String.format("%%%s = alloca %s", formalArg.name(), TypeDecider.llvmType(formalArg.type()));
+		addLine(s);
+	}
+
+	@Override
+	public void visit(VarDecl varDecl) {
+		String s = String.format("%%%s = alloca %s", varDecl.name(), TypeDecider.llvmType(varDecl.type()));
+		addLine(s);
+	}
+
+	// IF and JUMPS
+	@Override
+	public void visit(BlockStatement blockStatement) {
+		builder.append("label" + lastLabelNumber++ + ":");
+		for (var s : blockStatement.statements()) {
+			builder.append("\n");
+			s.accept(this);
+		}
+	}
+
+	@Override
+	public void visit(IfStatement ifStatement) {
+		ifStatement.cond().accept(this);
+		addLine("br i1 %_" + (lastRegisterNumber - 1) + ", " + "label %label" + lastLabelNumber + ", label %label"
+				+ (lastLabelNumber + 1));
+		if (!(ifStatement.thencase() instanceof BlockStatement)) {
```

Comment:
Why does it need to be a BlockStatement?

Reply:
look at 5_if example in recitation 5

Reply:
I understand now. but the logic here is confusing, if it is a BlockStatement then when you call it the first thing that happens is it adds a new label, if it isn't a BlockStatement then inside you add a label and then go to the expr.
So I think it's better to remove the line that adds the label inside BlockStatement then you change the code here with:

builder.append(label ...);
ifStatement.thencase().accept(this);
builder.append(label ...);
ifStatement.elsecase().accept(this);

also it seems that it's not right to generate the label inside the BlockStatement